src/com/xilinx/rapidwright/rwroute/RWRoute.java

```diff
@@ -0,0 +1,1841 @@
+					}
+				}
+				connection.setSourceRnode(this.createAddRoutableNode(this.rnodeId, connection.getSource(), sourceINTNode, RoutableType.PINFEED_O));
+				connection.setDirect(false);
+				indirect++;
+				connection.computeHpwl();
+				this.addConnectionSpanInfo(connection);
+			}
+		}
+		
+		if(indirect > 0) {
+			netWrapper.computeHPWLAndCenterCoordinates(boundingBoxExtensionX);
+			if(this.config.isUseBoundingBox()) {
+				for(Connection connection : netWrapper.getConnections()) {
+					if(connection.isDirect()) continue;
+					connection.computeConnectionBoundingBox(boundingBoxExtensionX, boundingBoxExtensionY,multiSLR);
+				}
+			}
+		}
+		return netWrapper;
+	}
+	
+	/**
+	 * Adds span info of a connection.
+	 * @param connection A connection of which span info is to be added.
+	 */
+	private void addConnectionSpanInfo(Connection connection) {
+		Integer counter = this.connectionSpan.get(connection.getHpwl());
+		if(counter == null) {
+			counter = 1;
+		}else {
+			counter += 1;
+		}
+		this.connectionSpan.put(connection.getHpwl(), counter);
+	}
+	
+	/**
+	 * Adds preserved nodes.
+	 * @param nodes A set of nodes to be preserved.
+	 * @param netToPreserve The net that uses those nodes.
+	 */
+	private void addPreservedNode(Set<Node> nodes, Net netToPreserve) {
+		for(Node node : nodes) {
+			this.addReservedNode(node, netToPreserve);
+		}
+	}
+	
+	protected void addReservedNode(Node node, Net netToPreserve) {
+		Net reserved = this.preservedNodes.get(node);
+		if(reserved == null) {
+			this.preservedNodes.put(node, netToPreserve);
+		}else if(!reserved.getName().equals(netToPreserve.getName())){
+			EDIFNet reservedLogical = reserved.getLogicalNet();
+			EDIFNet toReserveLogical = netToPreserve.getLogicalNet();
+			if(reservedLogical != null && toReserveLogical != null) {
+				if(!toReserveLogical.equals(reservedLogical))
+					System.out.println("WARNING: Conflicting node " + node + ":"); 
+					System.out.println("         " + netToPreserve.getName() + " \n         " + reserved.getName());
+			}else {
+				System.out.println("WARNING: Conflicting node " + node + ":"); 
+				System.out.println("         " + netToPreserve.getName() + " \n         " + reserved.getName());;
+			}
+		}	
+	}
+	
+	/**
+	 * Creates a {@link RoutableNode} Object based on a {@link Node} instance and avoids duplicates,
+	 * used for creating the source and sink rnodes of {@link Connection} instances.
+	 * @param rnodeGlobalIndex The design-wise index of created {@link RoutableNode} instances.
+	 * @param sitePinInst The source or sink {@link SitePinInst} instance.
+	 * @param node The node associated to the {@link SitePinInst} instance.
+	 * @param type The {@link RoutableType} of the {@link RoutableNode} Object.
+	 * @return The created {@link RoutableNode} instance.
+	 */
+	private Routable createAddRoutableNode(int rnodeGlobalIndex, SitePinInst sitePinInst, Node node, RoutableType type){
```

Comment:
Does `rnodeGlobalIndex` need to be passed in since it's always `this.rnodeId`?

Reply:
Not really. Removed.